### DISCUSSION TITLE
Experimental feature: Reduce Indexing Memory Usage

### DISCUSSION THREAD

**Discussion by Kerollmops** [meilisearch/product#652] [Experimental features]
Hey everyone 👋

For the v1.2 of Meilisearch (which will be released on 5th June), we will introduce an experimental feature that should help reduce the memory usage of the engine as an **experimental feature**. 

## Feature Abstract

 This flag enables the `MDB_WRITEMAP` option of LMDB, making the internal key-value store use much less RAM than usual. Unfortunately, it can reduce the write speed of it and therefore slow down the engine. You can [read more about this LMDB option](https://github.com/LMDB/lmdb/blob/3947014aed7ffe39a79991fa7fb5b234da47ad1a/libraries/liblmdb/lmdb.h#L573-L581) in the _lmdb.h_ header file.

This flag cannot be enabled on Windows. This is due to a different way of handling memory maps on this system than on others i.e. Unix-like.

### How to use the feature?
- Launch meilisearch using the `--experimental-reduce-indexing-memory-usage` CLI flag or the `MEILI_EXPERIMENTAL_REDUCE_INDEXING_MEMORY_USAGE` environment variable. This option does not accept any values.
- Index documents into memory and notice less RAM used by Meilisearch than without the flag.

---

## What is an experimental feature

- The API and the behavior of an experimental feature can break between two minor versions of Meilisearch.
- The whole experimental feature process can break between 2 minor versions of Meilisearch.
- The newly introduced experimental features or improvements can be removed in the future. Providing a feature as “experimental” does not guarantee it will be stable one day.
- The experimental features are supposed to be unstable regarding the API usage but **not bugged**.

⚠️ Experimental features are not recommended to be used in production

## Why is this feature not stable yet?

We have yet to determine if a lot impacts performance and whether the RAM usage reduction is significant enough. Could you give us your feedback on whether we want to ship it?

## When will the feature potentially be stable?

We have yet to determine when this feature could be stable since we need to iterate on it. As explained previously, we do not guarantee it will be stable someday.

Depending on the feedback we get, we plan to iterate on the feature every 1 or 2 releases (Currently, our release cycle is two months long).

🗣️  **You are welcome to give feedback about the feature or ask any questions; we can guide you through this new feature usage!**

**Comment by Zabrane** [comment]
Hi @Kerollmops 
What's the status of this feature (experimental_reduce_indexing_memory_usage)?
Can we safely turn it on forever?
We're always using the latest `Meilisearch`. Now on `1.10.0`.

**Reply by macraig** [reply]
Hi @Zabrane , the feature is still experimental, meaning we don't recommend using it in production environments. At the moment, we don't have a date for stabilization, but we’d appreciate any feedback you might have. How’s it been working for you? Any insights you can share about its impact on performance and whether the RAM usage reduction is significant enough would be really helpful.

**Reply by Zabrane** [reply]
> Hi @Zabrane , the feature is still experimental, meaning we don't recommend using it in production environments. At the moment, we don't have a date for stabilization, but we’d appreciate any feedback you might have. How’s it been working for you? Any insights you can share about its impact on performance and whether the RAM usage reduction is significant enough would be really helpful.

@macraig Ah I see. We will turn this experimental feature ON in a small cluster and I will report here my findings.

**Reply by macraig** [reply]
That would be great @Zabrane ! Looking forward to your findings!

**Comment by Averyy** [comment]
Been using Meilisearch quite happily for running a personal MCP server. Was using an old version for a while 1.9 and upgraded to v1.33 and ran into some memory issues. With this flag on a memory-constrained VPS (indexing 335K markdown documents).

Results with 2GB limit:
v1.9: Completed the full index in ~3 hours
v1.33 without flag: Out of memory/crashed at 110K docs (33%)
v1.33 with flag: Out of memory/crashed at 280K docs (83%)

With 3GB limit + flag: Completed the full index in 52 minutes.

The flag made a big difference: Instead of climbing until it dies, memory now cycles between 1.5-2.8GB and releases between operations rather than accumulating. Rate dropped from ~11K/min to ~6.5K/min as the index grew.

One thing that surprised me: smaller batches didn't help memory usage. 5K batches actually used more memory than 10K at the same document counts. I'm guessing more merge cycles = higher cumulative pressure? Counterintuitive but consistent across my tests.
Anyway, this flag let me run on a 4GB VPS that would've OOM'd otherwise. I'm currently using it without issues and didn't see anyone commenting on this thread so thought I would chime in. 

Appreciate you adding the flag, as it lets me save a lot of costs using a cheap VPS to host my personal MCP project.
